### PR TITLE
Add 2b embedding op

### PR DESCRIPTION
**Pull request by metascroy**

### CODE REVIEW

File: kernels/quantized/test/op_embedding2b_test.cpp [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+      weight_scales,
+      weight_zero_points,
+      quant_min,
+      quant_max,
+      indices,
+      out);
+
+  EXPECT_TENSOR_EQ(out, expected);
+
+  // Groupwise quantization. groupsize = 2
```
Comment (metascroy) [inline kernels/quantized/test/op_embedding2b_test.cpp:78]: These tests need to be fixed

File: exir/passes/_quant_patterns_and_replacements.py [hunk 2]
```diff
@@ -172,6 +172,138 @@ def embedding_byte_dtype_out_meta(
...
+    weight = torch.ops.quantized_decomposed.dequantize_per_channel_group.default(
+        weight,
+        weight_scales,
+        weight_zero_points,
+        weight_quant_min,
+        weight_quant_max,
+        weight.dtype,
+        group_size,
+        dtype,
+    )
```
Comment (kimishpatel) [inline exir/passes/_quant_patterns_and_replacements.py:281]: refactor this out to be shared with the one above

File: kernels/quantized/cpu/op_embedding2b.cpp [hunk 3]
```diff
@@ -0,0 +1,350 @@
+/*
```
Comment (kimishpatel) [inline kernels/quantized/cpu/op_embedding2b.cpp:1]: rename op_embedding4b.cpp to op_embedding_nbit.cpp and have 2b/4b in the same file. I suspect there is a lot of code reuse opportunity

File: kernels/quantized/cpu/op_embedding2b.cpp [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+      weight_quant_max);
+
+  if (out_dtype.has_value()) {
+    ET_CHECK_MSG(
+        out.scalar_type() == out_dtype.value(),
+        "output_dtype must match the dtype of the out tensor");
+  }
+}
+
+static inline int32_t weight_value(const unsigned char* w_data, int32_t index) {
```
Comment (kimishpatel) [inline kernels/quantized/cpu/op_embedding2b.cpp:124]: Like all you really need is for this function to behave differently between 2 and 4 bit. Rest should remain the same of unoptimized impl

**Review by kimishpatel — CHANGES_REQUESTED** [review]
Didnt do detail review because I see ton of code reuse opportunity. So lets fix that first.

File: exir/passes/_quant_patterns_and_replacements.py [hunk 5]
```diff
@@ -172,6 +172,138 @@ def embedding_byte_dtype_out_meta(
     )
 
 
+quantized_decomposed_lib.define(
```
[the remark this answers is not in the document]
Comment (metascroy) [reply exir/passes/_quant_patterns_and_replacements.py:175]: We can, although I planned to do something like that with the torchao version and keep this PR relatively uninvolved.

**Comment by metascroy** [comment]
> Didnt do detail review because I see ton of code reuse opportunity. So lets fix that first.

@kimishpatel for code reuse, are you OK with me getting rid of quantized_embedding_4bit_out and replacing it with quantized_embedding_xbit_out and adding an nbit arg to the signature?  Or would there be an BC concerns there?

**Comment by kimishpatel** [comment]
> > Didnt do detail review because I see ton of code reuse opportunity. So lets fix that first.
> 
> @kimishpatel for code reuse, are you OK with me getting rid of quantized_embedding_4bit_out and replacing it with quantized_embedding_xbit_out and adding an nbit arg to the signature? Or would there be an BC concerns there?
> 
> I guess your above comment suggests you are OK with that, but you also said "eventually deprecate even the _4bit one" so I wasn't sure if removing 4-bit in this PR is what you had in mind.

So my preference would be a) code refactor to make reuse easier b) have quantized_embedding_xbit_out but dont remove quantized_embedding_4bit_out quite yet.

File: kernels/quantized/test/op_embedding2b_test.cpp [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+  int64_t quant_min = -2;
+  int64_t quant_max = 1;
+
+  Tensor weight_scales = tf.make({4}, {0.5, 1.0, 1.5, 3.3});
+  Tensor weight_zero_points = tf.make({4}, {1, -2, 1, 0});
+  Tensor qweight = tfb.make({3, 1}, {236, 134, 228});
+  Tensor indices = tfl.make({3}, {0, 2, 1});
+  Tensor out = tf.zeros({3, 4});
+
+  ET_EXPECT_DEATH(
```
Comment (kimishpatel) [inline kernels/quantized/test/op_embedding2b_test.cpp:122]: Add a comment as to why this expects death

File: kernels/quantized/test/op_embedding2b_test.cpp [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+  TensorFactory<ScalarType::Long> tfl;
+
+  int64_t quant_min = -2;
+  int64_t quant_max = 1;
+
+  Tensor weight_scales = tf.make({2}, {0.5, 1.0});
+  Tensor weight_zero_points = tf.make({2}, {1, -2});
+  Tensor qweight = tfb.make({3, 1}, {236, 134, 228});
+  Tensor indices = tfl.make({3}, {0, 2, 1});
+  Tensor out = tf.zeros({3, 4});
```
Comment (kimishpatel) [inline kernels/quantized/test/op_embedding2b_test.cpp:147]: Do add a test where # of elements dont nicely pack. I presume you probably will need to throw

**Review by kimishpatel — APPROVED** [review]
a nit and a request for another test to be added.

File: kernels/quantized/test/op_embedding2b_test.cpp (hunk 7 above)
Comment (metascroy) [reply kernels/quantized/test/op_embedding2b_test.cpp:148]: Added 3rd death test

**Pull request closed without merging**